torch/csrc/jit/codegen/cuda/lower_alias_memory.cpp

```diff
@@ -61,127 +63,903 @@ class SymbolicSizePrinter : private kir::IrVisitor {
+    //  ite but could fill in the blank here
+    //  if this printer can be used for
+    //  other passes or we have more
+    //  complex ite pattern.
+    TORCH_INTERNAL_ASSERT(false, "unsupported");
+  }
+
+  void printAllocInfo(const kir::Allocate* alloc);
+
+  std::stringstream& indent() {
+    for (int i = 0; i < indent_level_; i++) {
+      os_ << "  ";
+    }
+    return os_;
+  }
+
+  void startBlock() {
+    indent_level_++;
+  }
+
+  void endBlock() {
+    indent_level_--;
+  }
+
+ private:
+  std::stringstream os_;
+  kir::IrPrinter ir_printer_;
+  int indent_level_ = 0;
+
+  std::vector<DebugEntryPtr> debug_info_;
+  BufferUseDefInfo* buffer_info_ = nullptr;
+};
+
+//! Utility class for modeling the liveness interval.
+//! The first write and last read
+//! is based on the position on the linear order within
+//! the Kernel IR.
+//!  The interval is semi-open,
+//!     i.e. [First_Write, Last_Read)
+//!  So the buffer is NOT available at exactly First_Write
+//!   position while it IS available at Last_Read.
+class BufferLiveInterval {
+ public:
+  // Simple detection of intersection of two intervals
+  bool intersect(BufferLiveInterval* other) {
+    if (first_write_pos_ <= other->first_write_pos_) {
+      return other->first_write_pos_ < last_read_pos_;
+    } else {
+      return first_write_pos_ < other->last_read_pos_;
+    }
+  }
+
+  void markWrite(int pos) {
+    if (first_write_pos_ == -1) {
+      first_write_pos_ = pos;
+    }
+  }
+
+  void markRead(int pos) {
+    last_read_pos_ = pos;
+    TORCH_INTERNAL_ASSERT(
+        pos > first_write_pos_, "marking a read before write");
+    all_read_pos_.push_back(pos);
+  }
+
+  const auto& allReads() {
+    return all_read_pos_;
+  }
+
+  auto firstWrite() const {
+    return first_write_pos_;
+  }
+
+  auto lastRead() const {
+    return last_read_pos_;
+  }
+
+  std::string toString() {
+    std::stringstream ss;
+    ss << "[ " << first_write_pos_ << " , " << last_read_pos_ << " )";
+    return ss.str();
+  }
+
+ private:
+  int first_write_pos_ = -1;
+  int last_read_pos_ = -1;
+  std::vector<int> all_read_pos_;
+};
+
+using BufferLiveIntervalPtrList = std::vector<BufferLiveInterval*>;
+
+//! Thin struct to keep track of loops. The actual loop body is
+//!  considered live in [start_pos, end_pos)
+struct ScopeInfo {
+  int start_pos = -1;
+  int end_pos = -1;
+
+  // nullptr means it's global scope
+  kir::ForLoop* loop = nullptr;
+};
+
+using ScopeInfoOwningPtr = std::unique_ptr<ScopeInfo>;
+using ScopeInfoOwningPtrList = std::vector<ScopeInfoOwningPtr>;
+
+//! Utility class to record the read and write of each
+//! allocated buffer.
 //!
-class AllocateReuseModifier {
-  // Alias local memory if it exceeds this threshold
-  static constexpr size_t kRegisterSizeThreshold = 1;
+//! Note:
+//!  this simplified interval analysis only works on pointwise ops and
+//!  reductions and broadcast. With no non-trivial IfThenElse and no
+//!  non-trivial re-computation.
```

Comment:
Actually this comment is slightly behind since there are already cases like this. By `non-trivial` re-computation I meant multiple issue of the same pixel in a pointwise op, for example:
```
for i = 0...10
  for j = 0...10
   a[i+j] = b[i+j]; // a few of b[...] gets visited multiple times by this line
```

So in this case line-number based liveness just doesn't tell much and would need a slightly more complex version of dependency test. It looks like most of the pointwise, reduction and most of broadcast do not meant to generate code like this. 

But Id's with `halo` can actually do this.

So current approach is just disable sharing any allocations with halo and we could maybe extend to more precise liveness analysis at some point if we need to support re-use with more patterns like this.